[Model Cleanup] Cleanup Model #ifdefs

weinig opened the pull request:
#### 22cf38e52362dfec24edcbeb57e472db7dd5b61d
<pre>
[Model Cleanup] Cleanup Model #ifdefs
<a href="https://bugs.webkit.org/show_bug.cgi?id=300587">https://bugs.webkit.org/show_bug.cgi?id=300587</a>

Reviewed by Mike Wyrzykowski.

As a first pass at cleaning up the model code, this replaces #ifdefs
like ENABLE(MODEL_PROCESS), which WebCore should not know about, with
feature specific macros. Specifically, we add:

    - ENABLE(MODEL_CONTEXT)
    - ENABLE(MODEL_ELEMENT_ACCESSIBILITY)
    - ENABLE(MODEL_ELEMENT_ANIMATIONS_CONTROL)
    - ENABLE(MODEL_ELEMENT_BOUNDING_BOX)
    - ENABLE(MODEL_ELEMENT_ENTITY_TRANSFORM)
    - ENABLE(MODEL_ELEMENT_ENVIRONMENT_MAP)
    - ENABLE(MODEL_ELEMENT_PORTAL)
    - ENABLE(MODEL_ELEMENT_STAGE_MODE)
    - ENABLE(MODEL_ELEMENT_STAGE_MODE_INTERACTION)

Additionally:
- The ModelPlayer::identifier() and ModelPlayerClient::didUnload() functions
  have been made unconditional, as they are universally implementable.
- The model accessibility children type has been abstracted into a new
  struct, ModelPlayerAccessibilityChildren to remove cocoa specific types
  from the model interfaces.
- ENABLE(GPUP_MODEL) has been renamed ENABLE(GPU_PROCESS_MODEL) to be
  consistent with other GPU Process enable macros.
- Fixed some sorting issues in wtf/Platform* headers.
- Added FIXMEs for improper use of ENALBE macros in PlatformHave.h
- Fixed define for HAVE_SCENEKIT back to 1.
- Moved ModelPlayerIdentifier out of platform and into Modules/model-element
  to be next to ModelPlayer and friends.

There remains more to do in follow ups:
- Remove usage of ENABLE(GPU_PROCESS_MODEL) in WebCore. This probably means
  moving all the DD prefixed code to WebKit.
- Add better runtime settings for use by HTMLModelElement.idl. It currently
  uses `EnabledBySetting=ModelProcessEnabled` which is improper.

* Source/WTF/wtf/PlatformEnable.h:
* Source/WTF/wtf/PlatformEnableCocoa.h:
* Source/WTF/wtf/PlatformHave.h:
* Source/WebCore/Headers.cmake:
* Source/WebCore/Modules/model-element/DDModelPlayer.h:
* Source/WebCore/Modules/model-element/DDModelPlayer.mm:
* Source/WebCore/Modules/model-element/HTMLModelElement.cpp:
* Source/WebCore/Modules/model-element/HTMLModelElement.h:
* Source/WebCore/Modules/model-element/HTMLModelElement.idl:
* Source/WebCore/Modules/model-element/ModelPlayer.cpp:
* Source/WebCore/Modules/model-element/ModelPlayer.h:
* Source/WebCore/Modules/model-element/ModelPlayerAccessibilityChildren.h: Added.
* Source/WebCore/Modules/model-element/ModelPlayerClient.h:
* Source/WebCore/Modules/model-element/ModelPlayerIdentifier.h: Renamed from Source/WebCore/platform/graphics/ModelPlayerIdentifier.h.
* Source/WebCore/Modules/model-element/ModelPlayerTransformState.h:
* Source/WebCore/Modules/model-element/PlaceholderModelPlayer.cpp:
* Source/WebCore/Modules/model-element/PlaceholderModelPlayer.h:
* Source/WebCore/Modules/model-element/dummy/DummyModelPlayer.cpp:
* Source/WebCore/Modules/model-element/dummy/DummyModelPlayer.h:
* Source/WebCore/Modules/model-element/scenekit/SceneKitModelPlayer.h:
* Source/WebCore/Modules/model-element/scenekit/SceneKitModelPlayer.mm:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/accessibility/AXCoreObject.h:
* Source/WebCore/accessibility/AccessibilityObject.cpp:
* Source/WebCore/accessibility/AccessibilityObject.h:
* Source/WebCore/accessibility/ios/WebAccessibilityObjectWrapperIOS.mm:
* Source/WebCore/accessibility/isolatedtree/AXIsolatedObject.cpp:
* Source/WebCore/accessibility/isolatedtree/AXIsolatedObject.h:
* Source/WebCore/accessibility/mac/WebAccessibilityObjectWrapperMac.mm:
* Source/WebCore/page/ChromeClient.h:
* Source/WebCore/page/DragController.cpp:
* Source/WebCore/page/EventHandler.h:
* Source/WebCore/page/StageModeOperations.h:
* Source/WebCore/page/ios/EventHandlerIOS.mm:
* Source/WebCore/platform/DragItem.h:
* Source/WebCore/platform/graphics/GraphicsLayer.h:
* Source/WebCore/platform/graphics/ca/GraphicsLayerCA.cpp:
* Source/WebCore/platform/graphics/ca/GraphicsLayerCA.h:
* Source/WebCore/rendering/RenderLayerBacking.cpp:
* Source/WebCore/rendering/RenderModel.cpp:
* Source/WebKit/GPUProcess/graphics/Model/ModelObjectHeap.cpp:
* Source/WebKit/GPUProcess/graphics/Model/ModelObjectHeap.h:
* Source/WebKit/GPUProcess/graphics/Model/RemoteDDMesh.cpp:
* Source/WebKit/GPUProcess/graphics/Model/RemoteDDMesh.h:
* Source/WebKit/GPUProcess/graphics/Model/RemoteDDMesh.messages.in:
* Source/WebKit/GPUProcess/graphics/WebGPU/RemoteGPU.cpp:
* Source/WebKit/GPUProcess/graphics/WebGPU/RemoteGPU.messages.in:
* Source/WebKit/ModelProcess/cocoa/ModelProcessModelPlayerProxy.h:
* Source/WebKit/ModelProcess/cocoa/ModelProcessModelPlayerProxy.mm:
* Source/WebKit/WebProcess/GPU/graphics/Model/RemoteDDMeshProxy.cpp:
* Source/WebKit/WebProcess/GPU/graphics/WebGPU/RemoteGPUProxy.cpp:
* Source/WebKit/WebProcess/Model/ARKitInlinePreviewModelPlayer.h:
* Source/WebKit/WebProcess/Model/ARKitInlinePreviewModelPlayer.mm:
* Source/WebKit/WebProcess/Model/ModelProcessModelPlayer.cpp:
* Source/WebKit/WebProcess/Model/ModelProcessModelPlayer.h:
* Source/WebKit/WebProcess/Model/WebModelPlayerProvider.cpp:

Canonical link: <a href="https://commits.webkit.org/301487@main">https://commits.webkit.org/301487@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/9bf36f03e3d14ac8a9d3031b3ea5ca207781c85b

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/126146 "3 style errors") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/45799 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/36602 "Built successfully") | [  ~~🛠 wpe~~](https://ews-build.webkit.org/#/builders/5/builds/132951 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 win~~](https://ews-build.webkit.org/#/builders/59/builds/77946 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/128017 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/46472 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/54344 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/5/builds/132951 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/59/builds/77946 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/129094 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/132/builds/37187 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/112847 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-wpe~~](https://ews-build.webkit.org/#/builders/5/builds/132951 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/133/builds/36087 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/135/builds/31029 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 wpe-cairo~~](https://ews-build.webkit.org/#/builders/65/builds/76425 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/118230 "Built successfully and passed tests") | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/106970 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/31252 "Passed tests") | [  ~~🛠 gtk~~](https://ews-build.webkit.org/#/builders/2/builds/135659 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/124617 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/52901 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/122/builds/40631 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/2/builds/135659 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/53360 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/109064 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/2/builds/135659 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/26566 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/49689 "Passed tests") | [  ~~🧪 mac-intel-wk2~~](https://ews-build.webkit.org/#/builders/137/builds/28032 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/50298 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/52798 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/58633 "Built successfully") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/157661 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/52126 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/39460 "Passed tests") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/55472 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/53836 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

mwyrzykowski reviewed (approved):
PR looks great but regarding the commit message, I'm not sure if it will be possible to remove all usages of `ENABLE(GPU_PROCESS_MODEL)` in WebCore.

We use WebGPU.framework to render the models and currently WebGPU.framework is only used from WebCore so not sure how we would bridge the WebCore gap.

The reason for this are as follows:

* WebGPU allows Swift C++ interop and the non-WebKit framework is in Swift
* eventually we want to possibly allow programmable shading for <model> and WGSL shaders would be required for security reasons
* WebGPU framework has the added security benefit of using ARC when working with Objective-C types

mwyrzykowski reviewed (changes requested):
I think this breaks the visionOS build:
```
In file included from /Volumes/Data/OpenSource/WebKitBuild/Release-xros/DerivedSources/WebCore/unified-sources/UnifiedSource70.cpp:2:
/Volumes/Data/OpenSource/Source/WebCore/Modules/model-element/HTMLModelElement.cpp:689:9: error: use of undeclared identifier 'supportsStageModeInteraction'
  689 |     if (supportsStageModeInteraction())
```

sorry I understand it is quite challenging to catch internal visionOS build errors externally.

mwyrzykowski commented on Source/WebCore/Modules/model-element/HTMLModelElement.cpp:
This is the line which is failing on visionOS it seems. 

But that also means `ENABLE(MODEL_ELEMENT_STAGE_MODE_INTERACTION)` is 0 on visionOS which seems unexpected?

weinig replied:
I found my bug. Uploading a new PR.

mwyrzykowski commented on Source/WTF/wtf/PlatformEnableCocoa.h:
```diff
@@ -582,6 +599,54 @@
+#endif
+
+#if !defined(ENABLE_MODEL_ELEMENT_STAGE_MODE_INTERACTION)
+#define ENABLE_MODEL_ELEMENT_STAGE_MODE ENABLE_MODEL_PROCESS
```
I think ^^ should be `ENABLE_MODEL_ELEMENT_STAGE_MODE_INTERACTION` and not `ENABLE_MODEL_ELEMENT_STAGE_MODE`

weinig replied:
Yup, that was it.

mwyrzykowski replied:
I'm just building internal visionOS with your change locally and once that completes I will approve 👍

mwyrzykowski reviewed (approved):
visionOS looks good now 👍

weinig committed:
[Model Cleanup] Cleanup Model #ifdefs
https://bugs.webkit.org/show_bug.cgi?id=300587

Reviewed by Mike Wyrzykowski.

As a first pass at cleaning up the model code, this replaces #ifdefs
like ENABLE(MODEL_PROCESS), which WebCore should not know about, with
feature specific macros. Specifically, we add:

    - ENABLE(MODEL_CONTEXT)
    - ENABLE(MODEL_ELEMENT_ACCESSIBILITY)
    - ENABLE(MODEL_ELEMENT_ANIMATIONS_CONTROL)
    - ENABLE(MODEL_ELEMENT_BOUNDING_BOX)
    - ENABLE(MODEL_ELEMENT_ENTITY_TRANSFORM)
    - ENABLE(MODEL_ELEMENT_ENVIRONMENT_MAP)
    - ENABLE(MODEL_ELEMENT_PORTAL)
    - ENABLE(MODEL_ELEMENT_STAGE_MODE)
    - ENABLE(MODEL_ELEMENT_STAGE_MODE_INTERACTION)

Additionally:
- The ModelPlayer::identifier() and ModelPlayerClient::didUnload() functions
  have been made unconditional, as they are universally implementable.
- The model accessibility children type has been abstracted into a new
  struct, ModelPlayerAccessibilityChildren to remove cocoa specific types
  from the model interfaces.
- ENABLE(GPUP_MODEL) has been renamed ENABLE(GPU_PROCESS_MODEL) to be
  consistent with other GPU Process enable macros.
- Fixed some sorting issues in wtf/Platform* headers.
- Added FIXMEs for improper use of ENALBE macros in PlatformHave.h
- Fixed define for HAVE_SCENEKIT back to 1.
- Moved ModelPlayerIdentifier out of platform and into Modules/model-element
  to be next to ModelPlayer and friends.

There remains more to do in follow ups:
- Remove usage of ENABLE(GPU_PROCESS_MODEL) in WebCore. This probably means
  moving all the DD prefixed code to WebKit.
- Add better runtime settings for use by HTMLModelElement.idl. It currently
  uses `EnabledBySetting=ModelProcessEnabled` which is improper.

* Source/WTF/wtf/PlatformEnable.h:
* Source/WTF/wtf/PlatformEnableCocoa.h:
* Source/WTF/wtf/PlatformHave.h:
* Source/WebCore/Headers.cmake:
* Source/WebCore/Modules/model-element/DDModelPlayer.h:
* Source/WebCore/Modules/model-element/DDModelPlayer.mm:
* Source/WebCore/Modules/model-element/HTMLModelElement.cpp:
* Source/WebCore/Modules/model-element/HTMLModelElement.h:
* Source/WebCore/Modules/model-element/HTMLModelElement.idl:
* Source/WebCore/Modules/model-element/ModelPlayer.cpp:
* Source/WebCore/Modules/model-element/ModelPlayer.h:
* Source/WebCore/Modules/model-element/ModelPlayerAccessibilityChildren.h: Added.
* Source/WebCore/Modules/model-element/ModelPlayerClient.h:
* Source/WebCore/Modules/model-element/ModelPlayerIdentifier.h: Renamed from Source/WebCore/platform/graphics/ModelPlayerIdentifier.h.
* Source/WebCore/Modules/model-element/ModelPlayerTransformState.h:
* Source/WebCore/Modules/model-element/PlaceholderModelPlayer.cpp:
* Source/WebCore/Modules/model-element/PlaceholderModelPlayer.h:
* Source/WebCore/Modules/model-element/dummy/DummyModelPlayer.cpp:
* Source/WebCore/Modules/model-element/dummy/DummyModelPlayer.h:
* Source/WebCore/Modules/model-element/scenekit/SceneKitModelPlayer.h:
* Source/WebCore/Modules/model-element/scenekit/SceneKitModelPlayer.mm:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/accessibility/AXCoreObject.h:
* Source/WebCore/accessibility/AccessibilityObject.cpp:
* Source/WebCore/accessibility/AccessibilityObject.h:
* Source/WebCore/accessibility/ios/WebAccessibilityObjectWrapperIOS.mm:
* Source/WebCore/accessibility/isolatedtree/AXIsolatedObject.cpp:
* Source/WebCore/accessibility/isolatedtree/AXIsolatedObject.h:
* Source/WebCore/accessibility/mac/WebAccessibilityObjectWrapperMac.mm:
* Source/WebCore/page/ChromeClient.h:
* Source/WebCore/page/DragController.cpp:
* Source/WebCore/page/EventHandler.h:
* Source/WebCore/page/StageModeOperations.h:
* Source/WebCore/page/ios/EventHandlerIOS.mm:
* Source/WebCore/platform/DragItem.h:
* Source/WebCore/platform/graphics/GraphicsLayer.h:
* Source/WebCore/platform/graphics/ca/GraphicsLayerCA.cpp:
* Source/WebCore/platform/graphics/ca/GraphicsLayerCA.h:
* Source/WebCore/rendering/RenderLayerBacking.cpp:
* Source/WebCore/rendering/RenderModel.cpp:
* Source/WebKit/GPUProcess/graphics/Model/ModelObjectHeap.cpp:
* Source/WebKit/GPUProcess/graphics/Model/ModelObjectHeap.h:
* Source/WebKit/GPUProcess/graphics/Model/RemoteDDMesh.cpp:
* Source/WebKit/GPUProcess/graphics/Model/RemoteDDMesh.h:
* Source/WebKit/GPUProcess/graphics/Model/RemoteDDMesh.messages.in:
* Source/WebKit/GPUProcess/graphics/WebGPU/RemoteGPU.cpp:
* Source/WebKit/GPUProcess/graphics/WebGPU/RemoteGPU.messages.in:
* Source/WebKit/ModelProcess/cocoa/ModelProcessModelPlayerProxy.h:
* Source/WebKit/ModelProcess/cocoa/ModelProcessModelPlayerProxy.mm:
* Source/WebKit/WebProcess/GPU/graphics/Model/RemoteDDMeshProxy.cpp:
* Source/WebKit/WebProcess/GPU/graphics/WebGPU/RemoteGPUProxy.cpp:
* Source/WebKit/WebProcess/Model/ARKitInlinePreviewModelPlayer.h:
* Source/WebKit/WebProcess/Model/ARKitInlinePreviewModelPlayer.mm:
* Source/WebKit/WebProcess/Model/ModelProcessModelPlayer.cpp:
* Source/WebKit/WebProcess/Model/ModelProcessModelPlayer.h:
* Source/WebKit/WebProcess/Model/WebModelPlayerProvider.cpp:

Canonical link: https://commits.webkit.org/301487@main
